// @pinetwork-js/nodepi/build/structures/ressources/stellar/Asset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Asset = void 0;
/**
 * Structure representing a Stellar Asset.
 */
class Asset {
    client;
    /**
     * The id of the asset.
     */
    id;
    /**
     * The asset type. Either `credit_alphanum4` or `credit_alphanum12`.
     */
    assetType;
    /**
     * The asset code.
     */
    assetCode;
    /**
     * The Stellar address of the asset issuer.
     */
    assetIssuerId;
    /**
     * A cursor value for use in pagination.
     */
    pagingToken;
    /**
     * The amount of authorized units issued for the asset.
     */
    amount;
    /**
     * The number of accounts that have issued a trustline to the asset. If the `authRequired` flag for the asset issuer is set to `true`, this number only includes the accounts who have both set up a trustline to the asset and have been authorized to hold the asset.
     */
    accountCount;
    /**
     * Denote the enabling/disabling of certain asset issuer privileges.
     */
    flags;
    constructor(client, data) {
        this.client = client;
        this.$patch(data);
    }
    $patch(data) {
        this.id = `${data.asset_code}:${data.asset_issuer}`;
        this.assetType = data.asset_type;
        this.assetCode = data.asset_code;
        this.assetIssuerId = data.asset_issuer;
        this.pagingToken = data.paging_token;
        this.amount = Number(data.amount);
        this.accountCount = data.num_accounts;
        this.flags = {
            authImmutable: data.flags.auth_immutable,
            authRequired: data.flags.auth_required,
            authRevocable: data.flags.auth_revocable,
        };
    }
    /**
     * Get the asset issuer account.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the asset issuer account instead.
     * @returns The account of the asset issuer.
     */
    getAssetIssuerAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.assetIssuerId, true, !forceUpdate);
    }
}
exports.Asset = Asset;
