// @pinetwork-js/nodepi/build/stellar/streams/EffectStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EffectStream = void 0;
const BaseStream_1 = require("./BaseStream");
/**
 * Stream manager for Stellar effects.
 */
class EffectStream extends BaseStream_1.BaseStream {
    stellar;
    constructor(stellar) {
        super(stellar);
        this.stellar = stellar;
    }
    /**
     * Handle new Stellar effects.
     *
     * @param effect - The received Stellar effect.
     */
    onMessage(effect) {
        const cachedEffect = this.stellar.effects.add(effect);
        this.stellar.emit("effect" /* Effect */, cachedEffect);
    }
    /**
     * Listen for new Stellar effects.
     */
    listen() {
        if (this._stop) {
            return;
        }
        let callBuilder = this.stellar.server.effects();
        if (this.stellar.account) {
            callBuilder = callBuilder.forAccount(this.stellar.account.publicKey());
        }
        this.stop = callBuilder.cursor('now').stream({
            onmessage: this.onMessage.bind(this),
            onerror: this.onError.bind(this),
        });
    }
}
exports.EffectStream = EffectStream;
