// @pinetwork-js/nodepi/build/stellar/streams/OfferStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OfferStream = void 0;
const BaseStream_1 = require("./BaseStream");
/**
 * Stream manager for Stellar offers.
 */
class OfferStream extends BaseStream_1.BaseStream {
    stellar;
    constructor(stellar) {
        super(stellar);
        this.stellar = stellar;
    }
    /**
     * Handle new Stellar offers.
     *
     * @param offer - The received Stellar offer.
     */
    onMessage(offer) {
        const cachedOffer = this.stellar.offers.add(offer);
        this.stellar.emit("offer" /* Offer */, cachedOffer);
    }
    /**
     * Listen for new Stellar offers.
     */
    listen() {
        if (this._stop) {
            return;
        }
        let callBuilder = this.stellar.server.offers();
        if (this.stellar.account) {
            callBuilder = callBuilder.forAccount(this.stellar.account.publicKey());
        }
        this.stop = callBuilder.cursor('now').stream({
            onmessage: this.onMessage.bind(this),
            onerror: this.onError.bind(this),
        });
    }
}
exports.OfferStream = OfferStream;
