// @pinetwork-js/nodepi/build/stellar/streams/OperationStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OperationStream = void 0;
const BaseStream_1 = require("./BaseStream");
/**
 * Stream manager for Stellar operations.
 */
class OperationStream extends BaseStream_1.BaseStream {
    stellar;
    constructor(stellar) {
        super(stellar);
        this.stellar = stellar;
    }
    /**
     * Handle new Stellar operations.
     *
     * @param operation - The received Stellar operation.
     */
    onMessage(operation) {
        const cachedOperation = this.stellar.operations.add(operation);
        this.stellar.emit("operation" /* Operation */, cachedOperation);
    }
    /**
     * Listen for new Stellar operations.
     */
    listen() {
        if (this._stop) {
            return;
        }
        let callBuilder = this.stellar.server.operations();
        if (this.stellar.account) {
            callBuilder = callBuilder.forAccount(this.stellar.account.publicKey());
        }
        this.stop = callBuilder.cursor('now').stream({
            onmessage: this.onMessage.bind(this),
            onerror: this.onError.bind(this),
        });
    }
}
exports.OperationStream = OperationStream;
