// @pinetwork-js/nodepi/build/stellar/streams/PaymentStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentStream = void 0;
const BaseStream_1 = require("./BaseStream");
/**
 * Stream manager for Stellar payment operations.
 */
class PaymentStream extends BaseStream_1.BaseStream {
    stellar;
    constructor(stellar) {
        super(stellar);
        this.stellar = stellar;
    }
    /**
     * Handle new Stellar payment operations.
     *
     * @param payment - The received Stellar payment operation.
     */
    onMessage(payment) {
        const cachedPayment = this.stellar.operations.add(payment);
        this.stellar.emit("payment" /* Payment */, cachedPayment);
    }
    /**
     * Listen for new Stellar payment operations.
     */
    listen() {
        if (this._stop) {
            return;
        }
        let callBuilder = this.stellar.server.payments();
        if (this.stellar.account) {
            callBuilder = callBuilder.forAccount(this.stellar.account.publicKey());
        }
        this.stop = callBuilder.cursor('now').stream({
            onmessage: this.onMessage.bind(this),
            onerror: this.onError.bind(this),
        });
    }
}
exports.PaymentStream = PaymentStream;
