// @pinetwork-js/nodepi/build/stellar/streams/TradeStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TradeStream = void 0;
const BaseStream_1 = require("./BaseStream");
/**
 * Stream manager for Stellar trades.
 */
class TradeStream extends BaseStream_1.BaseStream {
    stellar;
    constructor(stellar) {
        super(stellar);
        this.stellar = stellar;
    }
    /**
     * Handle new Stellar trades.
     *
     * @param trade - The received Stellar trade.
     */
    onMessage(trade) {
        const cachedTrade = this.stellar.trades.add(trade);
        this.stellar.emit("trade" /* Trade */, cachedTrade);
    }
    /**
     * Listen for new Stellar trades.
     */
    listen() {
        if (this._stop) {
            return;
        }
        let callBuilder = this.stellar.server.trades();
        if (this.stellar.account) {
            callBuilder = callBuilder.forAccount(this.stellar.account.publicKey());
        }
        this.stop = callBuilder.cursor('now').stream({
            onmessage: this.onMessage.bind(this),
            onerror: this.onError.bind(this),
        });
    }
}
exports.TradeStream = TradeStream;
