// @pinetwork-js/nodepi/build/stellar/streams/TransactionStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionStream = void 0;
const BaseStream_1 = require("./BaseStream");
/**
 * Stream manager for Stellar transactions.
 */
class TransactionStream extends BaseStream_1.BaseStream {
    stellar;
    constructor(stellar) {
        super(stellar);
        this.stellar = stellar;
    }
    /**
     * Handle new Stellar transactions.
     *
     * @param transaction - The received Stellar transaction.
     */
    onMessage(transaction) {
        const cachedTransaction = this.stellar.transactions.add(transaction);
        this.stellar.emit("transaction" /* Transaction */, cachedTransaction);
    }
    /**
     * Listen for new Stellar transactions.
     */
    listen() {
        if (this._stop) {
            return;
        }
        let callBuilder = this.stellar.server.transactions();
        if (this.stellar.account) {
            callBuilder = callBuilder.forAccount(this.stellar.account.publicKey());
        }
        this.stop = callBuilder.cursor('now').stream({
            onmessage: this.onMessage.bind(this),
            onerror: this.onError.bind(this),
        });
    }
}
exports.TransactionStream = TransactionStream;
