// @pinetwork-js/nodepi/build/structures/ressources/stellar/effects/AccountSponsorshipUpdatedEffect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountSponsorshipUpdatedEffect = void 0;
const Effect_1 = require("./Effect");
/**
 * Structure representing a Stellar Account Sponsorship Updated Effect.
 */
class AccountSponsorshipUpdatedEffect extends Effect_1.Effect {
    /**
     * The id of the former account sponsor.
     */
    formerSponsorId;
    /**
     * The id of the new account sponsor.
     */
    newSponsorId;
    constructor(client, data) {
        super(client, data);
        this.$patch(data);
    }
    $patch(data) {
        super.$patch(data);
        this.formerSponsorId = data.former_sponsor;
        this.newSponsorId = data.new_sponsor;
    }
    /**
     * Get the former sponsor of the account.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the former sponsor account instead.
     * @returns The account of the former sponsor.
     */
    getFormerSponsorAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.formerSponsorId, true, !forceUpdate);
    }
    /**
     * Get the new sponsor of the account.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the new sponsor account instead.
     * @returns The account of the new sponsor.
     */
    getNewSponsorAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.newSponsorId, true, !forceUpdate);
    }
}
exports.AccountSponsorshipUpdatedEffect = AccountSponsorshipUpdatedEffect;
