// @pinetwork-js/nodepi/build/structures/ressources/stellar/effects/DataSponsorshipCreatedEffect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataSponsorshipCreatedEffect = void 0;
const Effect_1 = require("./Effect");
/**
 * Structure representing a Stellar Data Sponsorship Created Effect.
 */
class DataSponsorshipCreatedEffect extends Effect_1.Effect {
    /**
     * The name of the data entry.
     */
    dataName;
    /**
     * The id of the new data sponsor.
     */
    sponsorId;
    constructor(client, data) {
        super(client, data);
        this.$patch(data);
    }
    $patch(data) {
        super.$patch(data);
        this.dataName = data.data_name;
        this.sponsorId = data.sponsor;
    }
    /**
     * Get the data entry of the account.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the data account entry instead.
     * @returns The data entry of the account.
     */
    async getDataAccountEntry(forceUpdate = false) {
        const account = await this.getAccount();
        return account.getDataEntry(this.dataName, forceUpdate);
    }
    /**
     * Get the new sponsor of the account.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the sponsor account instead.
     * @returns The account of the new sponsor.
     */
    getSponsorAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.sponsorId, true, !forceUpdate);
    }
}
exports.DataSponsorshipCreatedEffect = DataSponsorshipCreatedEffect;
