// @pinetwork-js/nodepi/build/structures/ressources/stellar/effects/Effect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Effect = void 0;
/**
 * Structure representing a Stellar Effect.
 */
class Effect {
    client;
    /**
     * The id of the effect.
     */
    id;
    /**
     * A cursor value for use in pagination.
     */
    pagingToken;
    /**
     * The name of the effect type.
     */
    typeName;
    /**
     * The numeric type of the effect.
     */
    type;
    /**
     * The account that originates the effect.
     */
    accountId;
    /**
     * The effect's creation date.
     */
    createdAt;
    /**
     * The effect's creation timestamp.
     */
    createdTimestamp;
    constructor(client, data) {
        this.client = client;
        this.$patch(data);
    }
    $patch(data) {
        this.id = data.id;
        this.pagingToken = data.paging_token;
        this.typeName = data.type;
        this.type = data.type_i;
        this.accountId = data.account;
        this.createdAt = new Date(data.created_at);
        this.createdTimestamp = this.createdAt.getTime();
    }
    /**
     * Get the account the effect belongs to.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the account instead.
     * @returns The account the effect belongs to.
     */
    getAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.accountId, true, !forceUpdate);
    }
}
exports.Effect = Effect;
