// @pinetwork-js/nodepi/build/structures/ressources/stellar/effects/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EffectTypeNames = void 0;
__exportStar(require("./AccountCreatedEffect"), exports);
__exportStar(require("./AccountCreditedEffect"), exports);
__exportStar(require("./AccountDebitedEffect"), exports);
__exportStar(require("./AccountFlagsUpdatedEffect"), exports);
__exportStar(require("./AccountHomeDomainUpdatedEffect"), exports);
__exportStar(require("./AccountSponsorshipCreatedEffect"), exports);
__exportStar(require("./AccountSponsorshipRemovedEffect"), exports);
__exportStar(require("./AccountSponsorshipUpdatedEffect"), exports);
__exportStar(require("./AccountThresholdsUpdatedEffect"), exports);
__exportStar(require("./ClaimableBalanceClaimantCreatedEffect"), exports);
__exportStar(require("./ClaimableBalanceClaimedEffect"), exports);
__exportStar(require("./ClaimableBalanceCreatedEffect"), exports);
__exportStar(require("./ClaimableBalanceSponsorshipCreatedEffect"), exports);
__exportStar(require("./ClaimableBalanceSponsorshipRemovedEffect"), exports);
__exportStar(require("./ClaimableBalanceSponsorshipUpdatedEffect"), exports);
__exportStar(require("./DataCreatedEffect"), exports);
__exportStar(require("./DataRemovedEffect"), exports);
__exportStar(require("./DataSponsorshipCreatedEffect"), exports);
__exportStar(require("./DataSponsorshipRemovedEffect"), exports);
__exportStar(require("./DataSponsorshipUpdatedEffect"), exports);
__exportStar(require("./DataUpdatedEffect"), exports);
__exportStar(require("./Effect"), exports);
__exportStar(require("./SequenceBumpedEffect"), exports);
__exportStar(require("./SignerCreatedEffect"), exports);
__exportStar(require("./SignerRemovedEffect"), exports);
__exportStar(require("./SignerSponsorshipCreatedEffect"), exports);
__exportStar(require("./SignerSponsorshipRemovedEffect"), exports);
__exportStar(require("./SignerSponsorshipUpdatedEffect"), exports);
__exportStar(require("./SignerUpdatedEffect"), exports);
__exportStar(require("./TradeEffect"), exports);
__exportStar(require("./TrustlineAuthorizedEffect"), exports);
__exportStar(require("./TrustlineAuthorizedToMaintainLiabilitiesEffect"), exports);
__exportStar(require("./TrustlineCreatedEffect"), exports);
__exportStar(require("./TrustlineDeauthorizedEffect"), exports);
__exportStar(require("./TrustlineFlagsUpdatedEffect"), exports);
__exportStar(require("./TrustlineRemovedEffect"), exports);
__exportStar(require("./TrustlineSponsorshipCreatedEffect"), exports);
__exportStar(require("./TrustlineSponsorshipRemovedEffect"), exports);
__exportStar(require("./TrustlineSponsorshipUpdatedEffect"), exports);
__exportStar(require("./TrustlineUpdatedEffect"), exports);
var EffectTypeNames;
(function (EffectTypeNames) {
    EffectTypeNames["accountCreated"] = "account_created";
    EffectTypeNames["accountCredited"] = "account_credited";
    EffectTypeNames["accountDebited"] = "account_debited";
    EffectTypeNames["accountThresholdsUpdated"] = "account_thresholds_updated";
    EffectTypeNames["accountHomeDomainUpdated"] = "account_home_domain_updated";
    EffectTypeNames["accountFlagsUpdated"] = "account_flags_updated";
    EffectTypeNames["signerCreated"] = "signer_created";
    EffectTypeNames["signerRemoved"] = "signer_removed";
    EffectTypeNames["signerUpdated"] = "signer_updated";
    EffectTypeNames["trustlineCreated"] = "trustline_created";
    EffectTypeNames["trustlineRemoved"] = "trustline_removed";
    EffectTypeNames["trustlineUpdated"] = "trustline_updated";
    EffectTypeNames["trustlineAuthorized"] = "trustline_authorized";
    EffectTypeNames["trustlineDeauthorized"] = "trustline_deauthorized";
    EffectTypeNames["trustlineAuthorizedToMaintainLiabilities"] = "trustline_authorized_to_maintain_liabilities";
    EffectTypeNames["trustlineFlagsUpdated"] = "trustline_flags_updated";
    EffectTypeNames["trade"] = "trade";
    EffectTypeNames["dataCreated"] = "data_created";
    EffectTypeNames["dataRemoved"] = "data_removed";
    EffectTypeNames["dataUpdated"] = "data_updated";
    EffectTypeNames["sequenceBumped"] = "sequence_bumped";
    EffectTypeNames["claimableBalanceCreated"] = "claimable_balance_created";
    EffectTypeNames["claimableBalanceClaimantCreated"] = "claimable_balance_claimant_created";
    EffectTypeNames["claimableBalanceClaimed"] = "claimable_balance_claimed";
    EffectTypeNames["accountSponsorshipCreated"] = "account_sponsorship_created";
    EffectTypeNames["accountSponsorshipUpdated"] = "account_sponsorship_updated";
    EffectTypeNames["accountSponsorshipRemoved"] = "account_sponsorship_removed";
    EffectTypeNames["trustlineSponsorshipCreated"] = "trustline_sponsorship_created";
    EffectTypeNames["trustlineSponsorshipUpdated"] = "trustline_sponsorship_updated";
    EffectTypeNames["trustlineSponsorshipRemoved"] = "trustline_sponsorship_removed";
    EffectTypeNames["dataSponsorshipCreated"] = "data_sponsorship_created";
    EffectTypeNames["dataSponsorshipUpdated"] = "data_sponsorship_updated";
    EffectTypeNames["dataSponsorshipRemoved"] = "data_sponsorship_removed";
    EffectTypeNames["claimableBalanceSponsorshipCreated"] = "claimable_balance_sponsorship_created";
    EffectTypeNames["claimableBalanceSponsorshipUpdated"] = "claimable_balance_sponsorship_updated";
    EffectTypeNames["claimableBalanceSponsorshipRemoved"] = "claimable_balance_sponsorship_removed";
    EffectTypeNames["signerSponsorshipCreated"] = "signer_sponsorship_created";
    EffectTypeNames["signerSponsorshipUpdated"] = "signer_sponsorship_updated";
    EffectTypeNames["signerSponsorshipRemoved"] = "signer_sponsorship_removed";
})(EffectTypeNames = exports.EffectTypeNames || (exports.EffectTypeNames = {}));
