// @pinetwork-js/nodepi/build/structures/ressources/stellar/effects/SignerSponsorshipCreatedEffect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignerSponsorshipCreatedEffect = void 0;
const Effect_1 = require("./Effect");
/**
 * Structure representing a Stellar Signer Sponsorship Created Effect.
 */
class SignerSponsorshipCreatedEffect extends Effect_1.Effect {
    /**
     * The id of the signer.
     */
    signerId;
    /**
     * The id of the new signer sponsor.
     */
    sponsorId;
    constructor(client, data) {
        super(client, data);
        this.$patch(data);
    }
    $patch(data) {
        super.$patch(data);
        this.signerId = data.signer;
        this.sponsorId = data.sponsor;
    }
    /**
     * Get the account of the signer.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the signer account instead.
     * @returns The account of the signer.
     */
    getSignerAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.signerId, true, !forceUpdate);
    }
    /**
     * Get the new sponsor of the account.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the sponsor account instead.
     * @returns The account of the new sponsor.
     */
    getSponsorAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.sponsorId, true, !forceUpdate);
    }
}
exports.SignerSponsorshipCreatedEffect = SignerSponsorshipCreatedEffect;
