// @pinetwork-js/nodepi/build/structures/ressources/stellar/effects/SignerSponsorshipRemovedEffect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignerSponsorshipRemovedEffect = void 0;
const Effect_1 = require("./Effect");
/**
 * Structure representing a Stellar Signer Sponsorship Removed Effect.
 */
class SignerSponsorshipRemovedEffect extends Effect_1.Effect {
    /**
     * The id of the signer.
     */
    signerId;
    /**
     * The id of the former signer sponsor.
     */
    formerSponsorId;
    constructor(client, data) {
        super(client, data);
        this.$patch(data);
    }
    $patch(data) {
        super.$patch(data);
        this.signerId = data.signer;
        this.formerSponsorId = data.former_sponsor;
    }
    /**
     * Get the account of the signer.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the signer account instead.
     * @returns The account of the signer.
     */
    getSignerAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.signerId, true, !forceUpdate);
    }
    /**
     * Get the former sponsor of the account.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the former sponsor account instead.
     * @returns The account of the former sponsor.
     */
    getFormerSponsorAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.formerSponsorId, true, !forceUpdate);
    }
}
exports.SignerSponsorshipRemovedEffect = SignerSponsorshipRemovedEffect;
