// @pinetwork-js/nodepi/build/structures/ressources/stellar/effects/TrustlineFlagsUpdatedEffect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrustlineFlagsUpdatedEffect = void 0;
const Effect_1 = require("./Effect");
/**
 * Structure representing a Stellar Trustline Flags Updated Effect.
 */
class TrustlineFlagsUpdatedEffect extends Effect_1.Effect {
    /**
     * The type of the created asset. Either `native`, `credit_alphanum4`, or `credit_alphanum12`.
     */
    assetType;
    /**
     * The code for the created asset. Appears if `assetType` is not native.
     */
    assetCode;
    /**
     * The Stellar address of the created asset’s issuer. Appears if `assetType` is not native.
     */
    assetIssuerId;
    /**
     * The id of the account being authorized.
     */
    trustorId;
    /**
     * If set to `true`, the trustor has been authorized to perform transactions with its credit.
     */
    authorizedFlag;
    /**
     * If set to `true`, the trustor has been authorized to maintain and reduce liabilities for its credit.
     */
    authorizedToMaintainLiabilitesFlag;
    constructor(client, data) {
        super(client, data);
        this.$patch(data);
    }
    $patch(data) {
        super.$patch(data);
        this.assetType = data.asset_type;
        this.assetCode = data.asset_code;
        this.assetIssuerId = data.asset_issuer;
        this.trustorId = data.trustor;
        this.authorizedFlag = data.authorized_flag;
        this.authorizedToMaintainLiabilitesFlag = data.authorized_to_maintain_liabilites_flag;
    }
    /**
     * Get the asset if it's not `native`.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the asset instead.
     * @returns The asset if it's not `native`.
     */
    async getAsset(forceUpdate = false) {
        if (this.assetType === 'native') {
            return;
        }
        const assetId = `${this.assetCode}:${this.accountId}`;
        return this.client.stellar.assets.fetch(assetId, true, !forceUpdate);
    }
    /**
     * Get the account of the trustor.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the trustor account instead.
     */
    getTrustorAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.trustorId, true, !forceUpdate);
    }
}
exports.TrustlineFlagsUpdatedEffect = TrustlineFlagsUpdatedEffect;
