// @pinetwork-js/nodepi/build/structures/ressources/stellar/effects/TrustlineRemovedEffect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrustlineRemovedEffect = void 0;
const Effect_1 = require("./Effect");
/**
 * Structure representing a Stellar Trustline Removed Effect.
 */
class TrustlineRemovedEffect extends Effect_1.Effect {
    /**
     * The type of the removed asset. Either `native`, `credit_alphanum4`, or `credit_alphanum12`.
     */
    assetType;
    /**
     * The code for the removed asset. Appears if `assetType` is not native.
     */
    assetCode;
    /**
     * The Stellar address of the removed asset’s issuer. Appears if `assetType` is not native.
     */
    assetIssuerId;
    /**
     * The maximum amount of the removed asset.
     */
    limit;
    constructor(client, data) {
        super(client, data);
        this.$patch(data);
    }
    $patch(data) {
        super.$patch(data);
        this.assetType = data.asset_type;
        this.assetCode = data.asset_code;
        this.assetIssuerId = data.asset_issuer;
        this.limit = Number(data.limit);
    }
}
exports.TrustlineRemovedEffect = TrustlineRemovedEffect;
