// @pinetwork-js/nodepi/build/structures/ressources/stellar/effects/TrustlineCreatedEffect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrustlineCreatedEffect = void 0;
const Effect_1 = require("./Effect");
/**
 * Structure representing a Stellar Trustline Created Effect.
 */
class TrustlineCreatedEffect extends Effect_1.Effect {
    /**
     * The type of the created asset. Either `native`, `credit_alphanum4`, or `credit_alphanum12`.
     */
    assetType;
    /**
     * The code for the created asset. Appears if `assetType` is not native.
     */
    assetCode;
    /**
     * The Stellar address of the created asset’s issuer. Appears if `assetType` is not native.
     */
    assetIssuerId;
    /**
     * The maximum amount of the created asset.
     */
    limit;
    constructor(client, data) {
        super(client, data);
        this.$patch(data);
    }
    $patch(data) {
        super.$patch(data);
        this.assetType = data.asset_type;
        this.assetCode = data.asset_code;
        this.assetIssuerId = data.asset_issuer;
        this.limit = Number(data.limit);
    }
    /**
     * Get the created asset if it's not `native`.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the asset instead.
     * @returns The created asset if it's not `native`.
     */
    async getAsset(forceUpdate = false) {
        if (this.assetType === 'native') {
            return;
        }
        const assetId = `${this.assetCode}:${this.assetIssuerId}`;
        return this.client.stellar.assets.fetch(assetId, true, !forceUpdate);
    }
}
exports.TrustlineCreatedEffect = TrustlineCreatedEffect;
