// @pinetwork-js/nodepi/build/structures/ressources/stellar/effects/TrustlineSponsorshipCreatedEffect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrustlineSponsorshipCreatedEffect = void 0;
const Effect_1 = require("./Effect");
/**
 * Structure representing a Stellar Trustline Sponsorship Created Effect.
 */
class TrustlineSponsorshipCreatedEffect extends Effect_1.Effect {
    /**
     * The asset of the trustline in the SEP-11 form `asset_code:issuing_address` or `native` (for Pi).
     */
    assetInfo;
    /**
     * The id of the new trustline sponsor.
     */
    sponsorId;
    constructor(client, data) {
        super(client, data);
        this.$patch(data);
    }
    $patch(data) {
        super.$patch(data);
        this.assetInfo = data.asset;
        this.sponsorId = data.sponsor;
    }
    /**
     * Get the asset of the trustline if it's not `native`.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the asset instead.
     * @returns The asset of the trustline if it's not `native`.
     */
    async getAsset(forceUpdate = false) {
        if (this.assetInfo === 'native') {
            return;
        }
        return this.client.stellar.assets.fetch(this.assetInfo, true, !forceUpdate);
    }
    /**
     * Get the new sponsor of the account.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the sponsor account instead.
     * @returns The account of the new sponsor.
     */
    getSponsorAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.sponsorId, true, !forceUpdate);
    }
}
exports.TrustlineSponsorshipCreatedEffect = TrustlineSponsorshipCreatedEffect;
