// @pinetwork-js/nodepi/build/stellar/StreamManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamManager = void 0;
const streams_1 = require("./streams");
class StreamManager {
    stellar;
    ledgers;
    transactions;
    operations;
    payments;
    effects;
    accounts;
    offers;
    trades;
    constructor(stellar) {
        this.stellar = stellar;
        this.ledgers = new streams_1.LedgerStream(stellar);
        this.transactions = new streams_1.TransactionStream(stellar);
        this.operations = new streams_1.OperationStream(stellar);
        this.payments = new streams_1.PaymentStream(stellar);
        this.effects = new streams_1.EffectStream(stellar);
        this.accounts = new streams_1.AccountStream(stellar);
        this.offers = new streams_1.OfferStream(stellar);
        this.trades = new streams_1.TradeStream(stellar);
    }
    listen(event) {
        if (event === 'all') {
            this.ledgers.listen();
            this.transactions.listen();
            this.operations.listen();
            this.payments.listen();
            this.effects.listen();
            this.accounts.listen();
            this.offers.listen();
            this.trades.listen();
        }
        else {
            this[event].listen();
        }
    }
    stop(event) {
        if (event === 'all') {
            this.ledgers.stop();
            this.transactions.stop();
            this.operations.stop();
            this.payments.stop();
            this.effects.stop();
            this.accounts.stop();
            this.offers.stop();
            this.trades.stop();
        }
        else {
            this[event].stop();
        }
    }
}
exports.StreamManager = StreamManager;
