// @pinetwork-js/nodepi/build/clients/Client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const lodash_merge_1 = __importDefault(require("lodash.merge"));
const http_1 = require("../http");
const PaymentManager_1 = require("../structures/managers/platform/PaymentManager");
const StellarClient_1 = require("./StellarClient");
const defaultOptions = {
    requestTimeout: 20_000,
    retryLimit: 2,
    http2: true,
    apiUrl: 'https://socialchain.app/',
};
/**
 * The main client for interacting with the Pi Network Platform API.
 */
class Client {
    token;
    /**
     * The http request manager.
     */
    http;
    /**
     * The client for interacting with the Pi Network Stellar API.
     */
    stellar;
    /**
     * A manager for the Pi Network payments.
     */
    payments;
    constructor(token, { http, stellar }) {
        this.token = token;
        this.http = new http_1.HttpManager(this, http ? lodash_merge_1.default(defaultOptions, http) : defaultOptions);
        this.stellar = new StellarClient_1.StellarClient(this, stellar);
        this.payments = new PaymentManager_1.PaymentManager(this);
    }
}
exports.Client = Client;
