// @pinetwork-js/nodepi/build/structures/ressources/stellar/operations/ClaimClaimableBalanceOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClaimClaimableBalanceOperation = void 0;
const Operation_1 = require("./Operation");
/**
 * Structure representing a Stellar Claim Claimable Balance Operation.
 */
class ClaimClaimableBalanceOperation extends Operation_1.Operation {
    /**
     * The id of the claimable balance.
     */
    balanceId;
    /**
     * The id of the account which claimed the balance.
     */
    claimantId;
    constructor(client, data) {
        super(client, data);
        this.$patch(data);
    }
    $patch(data) {
        super.$patch(data);
        this.balanceId = data.balance_id;
        this.claimantId = data.claimant;
    }
    /**
     * Get the claimed balance.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the claimable balance instead.
     * @returns The claimed balance.
     */
    getClaimableBalance(forceUpdate = false) {
        return this.client.stellar.claimableBalances.fetch(this.balanceId, true, !forceUpdate);
    }
    /**
     * Get the account which claimed the balance.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the claimant account instead.
     * @returns The account of the claimant.
     */
    getClaimantAccount(forceUpdate = false) {
        return this.client.stellar.claimableBalances.fetch(this.claimantId, true, !forceUpdate);
    }
}
exports.ClaimClaimableBalanceOperation = ClaimClaimableBalanceOperation;
