// @pinetwork-js/nodepi/build/structures/ressources/stellar/operations/PathPaymentSendOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PathPaymentSendOperation = void 0;
const managers_1 = require("../../../managers");
const Operation_1 = require("./Operation");
/**
 * Structure representing a Stellar Path Payment Send Operation.
 */
class PathPaymentSendOperation extends Operation_1.Operation {
    /**
     * The type of asset being sent. Either `native`, `credit_alphanum4`, or `credit_alphanum12`.
     */
    assetType;
    /**
     * The code for the asset being sent. Appears if `assetType` is not native.
     */
    assetCode;
    /**
     * The Stellar address of the issuer of the asset being sent. Appears if `assetType` is not native.
     */
    assetIssuerId;
    /**
     * The payment sender’s public key.
     */
    senderId;
    /**
     * The payment recipient’s public key.
     */
    recipientId;
    /**
     * The amount received designated in the source asset.
     */
    amount;
    /**
     * The intermediary assets that this path hops through.
     */
    path;
    /**
     * The amount sent designated in the source asset.
     */
    sourceAmount;
    /**
     * The minimum amount of destination asset expected to be received.
     */
    destinationMin;
    /**
     * The type for the source asset. Either `native`, `credit_alphanum4`, or `credit_alphanum12`.
     */
    sourceAssetType;
    /**
     * The code for the source asset. Appears if `sourceAssetType` is not native.
     */
    sourceAssetcode;
    /**
     * The Stellar address of the source asset’s issuer. Appears if `sourceAssetType` is not native.
     */
    sourceAssetIssuerId;
    constructor(client, data) {
        super(client, data);
        this.$patch(data);
    }
    $patch(data) {
        super.$patch(data);
        this.assetType = data.asset_type;
        this.assetCode = data.asset_code;
        this.assetIssuerId = data.asset_issuer;
        this.senderId = data.from;
        this.recipientId = data.to;
        this.amount = Number(data.amount);
        this.path = data.path.map((path) => ({
            assetCode: path.asset_code,
            assetIssuerId: path.asset_issuer,
            assetType: path.asset_type,
        }));
        this.sourceAmount = Number(data.source_amount);
        this.destinationMin = data.destination_min;
        this.sourceAssetcode = data.source_asset_code;
        this.sourceAssetIssuerId = data.source_asset_issuer;
        this.sourceAssetType = data.source_asset_type;
    }
    /**
     * Get the sent asset if it's not `native`.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the sent asset instead.
     * @returns The sent asset if it's not `native`.
     */
    async getSentAsset(forceUpdate = false) {
        if (this.assetType === 'native') {
            return;
        }
        const assetId = `${this.assetCode}:${this.assetIssuerId}`;
        return this.client.stellar.assets.fetch(assetId, true, !forceUpdate);
    }
    /**
     * Get the source asset if it's not `native`.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the source asset instead.
     * @returns The source asset if it's not `native`.
     */
    async getSourceAsset(forceUpdate = false) {
        if (this.sourceAssetType === 'native') {
            return;
        }
        const assetId = `${this.sourceAssetcode}:${this.sourceAssetIssuerId}`;
        return this.client.stellar.assets.fetch(assetId, true, !forceUpdate);
    }
    /**
     * Get the non-native intermediary assets.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the intermediary assets asset instead.
     * @returns A manager for the intermediary assets.
     */
    async getIntermediaryAssets(forceUpdate = false) {
        const assets = await Promise.all(this.path
            .filter((asset) => asset.assetType !== 'native')
            .map((asset) => {
            const assetId = `${asset.assetCode}:${asset.assetIssuerId}`;
            return this.client.stellar.assets.fetch(assetId, true, !forceUpdate);
        }));
        return new managers_1.AssetManager(this.client.stellar, assets);
    }
    /**
     * Get the account of the payment sender.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the sender account instead.
     * @returns The account of the sender.
     */
    getSenderAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.senderId, true, !forceUpdate);
    }
    /**
     * Get the account of the payment recipient.
     *
     * @param forceUpdate - If set to `true`, it will not check in the cache and directly make a request to retrieve the recipient account instead.
     * @returns The account of the recipient.
     */
    getRecipientAccount(forceUpdate = false) {
        return this.client.stellar.accounts.fetch(this.recipientId, true, !forceUpdate);
    }
}
exports.PathPaymentSendOperation = PathPaymentSendOperation;
